junifer/storage/hdf5.py

```diff
@@ -0,0 +1,959 @@
+                )
+            )
+        # Parameter check pass; read metadata
+        metadata = read_hdf5(
+            fname=str(self.uri.resolve()),  # type: ignore
+            title="meta",
+            slash="ignore",
+        )
+        # Initialize MD5 variable
+        md5: str = ""
+
+        # Consider feature_md5
+        if feature_md5:
+            logger.debug(
+                f"Validating feature MD5 '{feature_md5}' in metadata "
+                f"for: {self.uri.resolve()} ..."  # type: ignore
+            )
+            # Validate MD5
+            if feature_md5 in metadata:
+                md5 = feature_md5  # type: ignore
+            else:
+                raise_error(msg=f"Feature MD5 '{feature_md5}' not found")
+
+        # Consider feature_name
+        elif feature_name:
+            logger.debug(
+                f"Validating feature name '{feature_name}' in metadata "
+                f"for: {self.uri.resolve()} ..."  # type: ignore
+            )
+            # Retrieve MD5 for feature_name
+            # Implicit counter for duplicate feature_name with different
+            # MD5; happens when something is wrong with marker computation
+            feature_name_duplicates_with_different_md5 = []
+            for md5, meta in metadata.items():
+                if meta["name"] == feature_name:
+                    feature_name_duplicates_with_different_md5.append(md5)
+
+            # Check for no / duplicate feature_name
+            if len(feature_name_duplicates_with_different_md5) == 0:
+                raise_error(msg=f"Feature '{feature_name}' not found")
+            elif len(feature_name_duplicates_with_different_md5) > 1:
+                raise_error(
+                    msg=(
+                        "More than one feature with name "
+                        f"'{feature_name}' found. You can bypass this "
+                        "issue by specifying a `feature_md5`."
+                    )
+                )
+
+            md5 = feature_name_duplicates_with_different_md5[0]
+
+        # Read data from HDF5
+        hdf_data = read_hdf5(
+            fname=str(self.uri.resolve()),  # type: ignore
+            title=md5,
+            slash="ignore",
+        )
+        reshaped_data = hdf_data["data"]
+
+        # Generate index for the data
+        logger.debug(f"Generating pandas.MultiIndex for {md5} ...")
+        # Create dictionary for aggregating index data
+        element_idx_dict = defaultdict(list)
+
+        if hdf_data["kind"] == "matrix":
+            # Get row and column count
+            n_rows, n_cols = hdf_data["data"][:, :, 0].shape
+            # Set rows for the index
+            for element in hdf_data["element"]:
+                for key, val in element.items():
+                    # Only one row per element
+                    element_idx_dict[key].append(val)
+            # Set column headers for dataframe
+            columns = list(
```

Comment:
Check what we do with the SQLIteFeatureStorage. It's not that simple as creating the product, because we might only want to to store a triangular matrix.

Indeed, the code from [here](https://github.com/juaml/junifer/blob/7acc0cd188d44bce4248504ce498e6d55256f945/junifer/storage/sqlite.py#L434) can be extracted to a function in storage.utils so an upper/lower/full matrix can be converted to row in the same way, in both places.

Reply:
Addressed with latest commits.